_posts/2018-06-21-vision-for-rust-and-wasm.md

```diff
@@ -0,0 +1,255 @@
+Everything we build is for naught if people can't learn how to use it
+themselves. So we are writing [The Rust and WebAssembly Book][book].
+
+Right now, it already has a lot of great content:
+
+* Getting up and running
+* Designing and implementing a non-trivial example (the Game of Life) that
+  integrates Rust and JavaScript
+* Tips for debugging, time profiling, and code size profiling
+* How to publish to npm with `wasm-pack`
+
+But it doesn't have a lot of continuity. It feels like a collection of
+appendices and random tutorials. We will have a follow up blog post detailing
+its specific needs, and how to help if you're interested.
+
+### wasm-bindgen
+
+[`wasm-bindgen` facilitates communication between Rust and
+JavaScript.][wasm-bindgen] You can import JavaScript things into Rust, and
+export Rust things to JavaScript. It allows you to send rich types like strings
+and structs between wasm and JavaScript, rather than only the simple integers
+and floats defined by the WebAssembly standard.
+
+Here is "Hello, World!" between Rust and JavaScript with `wasm-bindgen`. First,
+we import the `alert` function into Rust and export the `greet` function to
+JavaScript:
+
+```rust
+extern crate wasm_bindgen;
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+extern {
+    fn alert(s: &str);
+}
+
+#[wasm_bindgen]
+pub fn greet(name: &str) {
+    alert(&format!("Hello, {}!", name));
+}
+```
+
+Then, we import the wasm as an ECMAScript module in JavaScript, and call the
+`greet` function:
+
+```js
+import { greet } from "./hello_world";
+
+greet("World!");
+```
+
+<a href="/images/wasm-bindgen-architecture-current.png">
+  <img src="/images/wasm-bindgen-architecture-current.png" alt="wasm-bindgen architecture" style="float:right;width:50%;min-width:200px;max-width:400px;margin:1em"/>
+</a>
+
+How does `wasm-bindgen` work? Simplifying a bit, it is a procedural macro that
+takes in Rust source code annotated with `#[wasm_bindgen]` attributes,
+constructs an abstract syntax tree (AST), and then it emits two artifacts:
+
+1. Rust bindings that import JavaScript things and export Rust things.
+
+2. JavaScript bindings that expose a nice interface to Rust-exported things to
+   other JavaScript code and provide the Rust's desired imports.
+
+`wasm-bindgen`'s approach to JavaScript binding allows you to pay only for the
+imports that you use. Just because you imported the `window.alert` function, you
+don't end up with glue code for `window.document`.
+
+The big downside is that, right now, you always have to declare imports
+yourself. There are common imports for JavaScript functions and types and the
+Web platform APIs that will undoubtedly be repeated by many people many times
+over. Importing these by-hand is both boring and mechanical. We have a plan for
+fixing this, but you'll have to wait for a follow up blog post to learn more.
+
+<div style="clear: both"/>
+
+### wasm-pack
+
+[`wasm-pack` seeks to be a one-stop shop for building, optimizing, and
```

Comment:
we may add something here about how it runs the cli bindgen step for you.

Reply:
tho i guess that's below. anyways, it may be useful to reiterate it since it seems to be a common confusion, but i defer to whatever you think is best